### ISSUE TITLE
ElseIf operator incorrect disassembling the code.

### ISSUE THREAD

**Issue by Moorre** [acpica/acpica#131]
Hello!

I think it is some bug with ElseIf operator.

Example:

`Store (B1PR, Local0)`
`ShiftLeft (Local0, 0x08, Local1)`
`And (Local1, 0xFF00, Local1)`
`ShiftRight (Local0, 0x08, Local0)`
`Or (Local0, Local1, Local0)`
`If (LEqual (Local0, 0xFFFF))`
` {`
`    Store (0xFFFFFFFF, Index (STAT, One))`
`}`
`ElseIf (LGreaterEqual (Local0, 0x8000))`
` {`
` XOr (Local0, 0xFFFF, Local0)`
 `   Increment (Local0)`
`}`

Should be like this, if I use the 4.0A version:

`                                Store (B1PR, Local0)`
`                                ShiftLeft (Local0, 0x08, Local1)`
`                                And (Local1, 0xFF00, Local1)`
 `                               ShiftRight (Local0, 0x08, Local0)`
`                                Or (Local0, Local1, Local0)`
`                                If (LEqual (Local0, 0xFFFF))`
`                                {`
                                `    Store (0xFFFFFFFF, Index (STAT, One))`
                           `     }`
                               ` Else`
                               ` {`
                                  `  If (LGreaterEqual (Local0, 0x8000))`
                                 `   {`
                                       ` XOr (Local0, 0xFFFF, Local0)`
                                       ` Increment (Local0)`
                                `    }`
                                `    Store (Local0, Index (STAT, One)) // Here. It was lost with ElseIf. `
`}`

Example code attached.
[Samsung_R522.zip](https://github.com/acpica/acpica/files/209682/Samsung_R522.zip)

*** Sorry, I cant do the correction, it is glitching with code braces :(

**Comment by zetalog** [comment]
Could you help to re-format your report according to this:
https://guides.github.com/features/mastering-markdown/
See "Syntax highlighting" section.

**Comment by Moorre** [comment]
*** Sorry, I cant do the correction, it is glitching with code braces :(

**Comment by RehabMan** [comment]
Many other bad translations in this file.  Looks like the code does not contemplate this construct:

```
If (cond1)
{
  M1()
}
else
{
    if (cond2)
    {
        M2()
    }
    M3()
}
```

In the case there, If...Else...If to If..ElseIf is not possible.

The disassembler seems to throw away the code following the If(..){M2()} block.

For example:
```
                        If (LOr (LEqual (PSIZ, Zero), LGreater (PSIZ, 0x04)))
                        {
                            Return (Package (0x01)
                            {
                                0x00010110
                            })
                        }
                        Else
                        {
                            If (LEqual (PSIZ, 0x02))
                            {
                                Name (VID2, Package (0x02)
                                {
                                    0x00010110, 
                                    0x00010100
                                })
                                If (And (Local0, One))
                                {
                                    Store (0x00010110, Index (VID2, PPTR))
                                    Increment (PPTR)
                                }

                                If (And (Local0, 0x02))
                                {
                                    Store (0x00010100, Index (VID2, PPTR))
                                    Increment (PPTR)
                                }

                                If (And (Local0, 0x04))
                                {
                                    Store (0x00010200, Index (VID2, PPTR))
                                    Increment (PPTR)
                                }

                                If (And (Local0, 0x08))
                                {
                                    Store (0x00010120, Index (VID2, PPTR))
                                    Increment (PPTR)
                                }

                                Return (VID2)
                            }

                            If (LEqual (PSIZ, 0x03))
                            {
                                Name (VID3, Package (0x03)
                                {
                                    0x00010110, 
                                    0x00010100, 
                                    0x00010200
                                })
                                If (And (Local0, One))
                                {
                                    Store (0x00010110, Index (VID3, PPTR))
                                    Increment (PPTR)
                                }

                                If (And (Local0, 0x02))
                                {
                                    Store (0x00010100, Index (VID3, PPTR))
                                    Increment (PPTR)
                                }

                                If (And (Local0, 0x04))
                                {
                                    Store (0x00010200, Index (VID3, PPTR))
                                    Increment (PPTR)
                                }

                                If (And (Local0, 0x08))
                                {
                                    Store (0x00010120, Index (VID3, PPTR))
                                    Increment (PPTR)
                                }

                                Return (VID3)
                            }

                            If (LEqual (PSIZ, 0x04))
                            {
                                Name (VID4, Package (0x04)
                                {
                                    0x00010110, 
                                    0x00010100, 
                                    0x00010200, 
                                    0x00010120
                                })
                                If (And (Local0, One))
                                {
                                    Store (0x00010110, Index (VID4, PPTR))
                                    Increment (PPTR)
                                }

                                If (And (Local0, 0x02))
                                {
                                    Store (0x00010100, Index (VID4, PPTR))
                                    Increment (PPTR)
                                }

                                If (And (Local0, 0x04))
                                {
                                    Store (0x00010200, Index (VID4, PPTR))
                                    Increment (PPTR)
                                }

                                If (And (Local0, 0x08))
                                {
                                    Store (0x00010120, Index (VID4, PPTR))
                                    Increment (PPTR)
                                }

                                Return (VID4)
                            }
```

To:
```
                        If (LOr (LEqual (PSIZ, Zero), LGreater (PSIZ, 0x04)))
                        {
                            Return (Package (0x01)
                            {
                                0x00010110
                            })
                        }
                        ElseIf (LEqual (PSIZ, 0x02))
                        {
                            Name (VID2, Package (0x02)
                            {
                                0x00010110, 
                                0x00010100
                            })
                            If (And (Local0, One))
                            {
                                Store (0x00010110, Index (VID2, PPTR))
                                Increment (PPTR)
                            }

                            If (And (Local0, 0x02))
                            {
                                Store (0x00010100, Index (VID2, PPTR))
                                Increment (PPTR)
                            }

                            If (And (Local0, 0x04))
                            {
                                Store (0x00010200, Index (VID2, PPTR))
                                Increment (PPTR)
                            }

                            If (And (Local0, 0x08))
                            {
                                Store (0x00010120, Index (VID2, PPTR))
                                Increment (PPTR)
                            }

                            Return (VID2)
                        }
```

And this:
```
                    Method (_DGS, 0, NotSerialized)  // _DGS: Display Graphics State
                    {
                        If (SCIP ())
                        {
                            Return (NDDS (DID1))
                        }
                        Else
                        {
                            If (And (NSTE, One))
                            {
                                Return (One)
                            }

                            Return (Zero)
                        }
                    }
```

To:
```
                    Method (_DGS, 0, NotSerialized)  // _DGS: Display Graphics State
                    {
                        If (SCIP ())
                        {
                            Return (NDDS (DID1))
                        }
                        ElseIf (And (NSTE, One))
                        {
                            Return (One)
                        }
                    }
```

And the example from the OP:
```
                                If (LEqual (Local0, 0xFFFF))
                                {
                                    Store (0xFFFFFFFF, Index (STAT, One))
                                }
                                Else
                                {
                                    If (LGreaterEqual (Local0, 0x8000))
                                    {
                                        XOr (Local0, 0xFFFF, Local0)
                                        Increment (Local0)
                                    }

                                    Store (Local0, Index (STAT, One))
                                }

                                Store (SECW (0x82, 0xA2, Zero), Local0)
```

To:
```
                                If (LEqual (Local0, 0xFFFF))
                                {
                                    Store (0xFFFFFFFF, Index (STAT, One))
                                }
                                ElseIf (LGreaterEqual (Local0, 0x8000))
                                {
                                    XOr (Local0, 0xFFFF, Local0)
                                    Increment (Local0)
                                }

                                Store (SECW (0x82, 0xA2, Zero), Local0)
```

And:
```
                                If (LEqual (Local0, 0xFFFF))
                                {
                                    Store (0xFFFFFFFF, Index (STAT, One))
                                }
                                Else
                                {
                                    If (LGreaterEqual (Local0, 0x8000))
                                    {
                                        XOr (Local0, 0xFFFF, Local0)
                                        Increment (Local0)
                                    }

                                    Store (Local0, Index (STAT, One))
                                }

                                Store (B1RA, Local0)
```
To:
```
                                If (LEqual (Local0, 0xFFFF))
                                {
                                    Store (0xFFFFFFFF, Index (STAT, One))
                                }
                                ElseIf (LGreaterEqual (Local0, 0x8000))
                                {
                                    XOr (Local0, 0xFFFF, Local0)
                                    Increment (Local0)
                                }

                                Store (B1RA, Local0)
```

And:
```
    Method (HKDS, 1, Serialized)
    {
        If (IGDS)
        {
            If (LEqual (Zero, And (0x03, DSEN)))
            {
                If (LEqual (TRAP (TRTI, Arg0), Zero))
                {
                    If (LNotEqual (CADL, PADL))
                    {
                        Store (CADL, PADL)
                        If (LOr (LGreater (OSYS, 0x07D0), LLess (OSYS, 0x07D6)))
                        {
                            Notify (\_SB.PCI0, Zero)
                        }
                        Else
                        {
                            Notify (\_SB.PCI0.GFX0, Zero)
                        }

                        Sleep (0x02EE)
                    }

                    Notify (\_SB.PCI0.GFX0, 0x80)
                }
            }

            If (LEqual (One, And (0x03, DSEN)))
            {
                If (LEqual (TRAP (TRTI, Increment (Arg0)), Zero))
                {
                    Notify (\_SB.PCI0.GFX0, 0x81)
                }
            }
        }
        Else
        {
            If (LEqual (Zero, DSEN))
            {
                Store (\_SB.STRP (0x30, Arg0), Local0)
                If (LEqual (Local0, Zero))
                {
                    If (LNotEqual (\_SB.CADL, \_SB.PADL))
                    {
                        Store (\_SB.CADL, \_SB.PADL)
                        If (LGreaterEqual (OSYS, 0x07D1))
                        {
                            Notify (\_SB.PCI0, Zero)
                        }
                        Else
                        {
                            Notify (\_SB.PCI0.P0P2.NVID, Zero)
                        }

                        Sleep (0x02EE)
                    }

                    Notify (\_SB.PCI0.P0P2.NVID, 0x80)
                }
            }

            If (LEqual (One, DSEN))
            {
                Store (Increment (Arg0), Local1)
                Store (\_SB.STRP (0x30, Local1), Local0)
                If (LEqual (Local0, Zero))
                {
                    Notify (\_SB.PCI0.P0P2.NVID, 0x81)
                }
            }
        }
    }
```

To:
```
    Method (HKDS, 1, Serialized)
    {
        If (IGDS)
        {
            If (LEqual (Zero, And (0x03, DSEN)))
            {
                If (LEqual (TRAP (TRTI, Arg0), Zero))
                {
                    If (LNotEqual (CADL, PADL))
                    {
                        Store (CADL, PADL)
                        If (LOr (LGreater (OSYS, 0x07D0), LLess (OSYS, 0x07D6)))
                        {
                            Notify (\_SB.PCI0, Zero)
                        }
                        Else
                        {
                            Notify (\_SB.PCI0.GFX0, Zero)
                        }

                        Sleep (0x02EE)
                    }

                    Notify (\_SB.PCI0.GFX0, 0x80)
                }
            }

            If (LEqual (One, And (0x03, DSEN)))
            {
                If (LEqual (TRAP (TRTI, Increment (Arg0)), Zero))
                {
                    Notify (\_SB.PCI0.GFX0, 0x81)
                }
            }
        }
        ElseIf (LEqual (Zero, DSEN))
        {
            Store (\_SB.STRP (0x30, Arg0), Local0)
            If (LEqual (Local0, Zero))
            {
                If (LNotEqual (\_SB.CADL, \_SB.PADL))
                {
                    Store (\_SB.CADL, \_SB.PADL)
                    If (LGreaterEqual (OSYS, 0x07D1))
                    {
                        Notify (\_SB.PCI0, Zero)
                    }
                    Else
                    {
                        Notify (\_SB.PCI0.P0P2.NVID, Zero)
                    }

                    Sleep (0x02EE)
                }

                Notify (\_SB.PCI0.P0P2.NVID, 0x80)
            }
        }
    }
```

There are a few more examples in this file...
Disassemble it with iasl51 and iasl61, the compare with diffmerge and you'll see.

**Comment by acpibob** [comment]
Ok, we’ll look at it

**Comment by RehabMan** [comment]
Thanks Bob... this is a bad one.  

I have temporarily disabled the ElseIf translation code in my fork: https://github.com/RehabMan/Intel-iasl/commit/2b8962a964e1c255932dd8350593f6c5c7720de0

**Comment by acpibob** [comment]
Reproduced. We’re on it.
Thanks,
Bob

**Comment by Moorre** [comment]
I think, that I can close this issue?

**Comment by acpibob** [comment]
I'll close it when it is actually released.

**Comment by RehabMan** [comment]
This fix is in the 20160422 build.

**Comment by acpibob** [comment]
yes, closing.
